flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java

```diff
@@ -124,23 +158,70 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
 
     @Override
     public void onNewResourcesAvailable() {
-        maybeRescale();
+        rescaleWhenCooldownPeriodIsOver();
     }
 
     @Override
     public void onNewResourceRequirements() {
-        maybeRescale();
+        rescaleWhenCooldownPeriodIsOver();
+    }
+
+    /** Force rescaling as long as the target parallelism is different from the current one. */
+    private void forceRescale() {
+        if (context.shouldRescale(getExecutionGraph(), true)) {
+            getLogger()
+                    .info(
+                            "Added resources are still there after {} time({}), force a rescale.",
+                            JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX.key(),
+                            scalingIntervalMax);
+            context.goToRestarting(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    Duration.ofMillis(0L),
+                    getFailures());
+        }
     }
 
+    /**
+     * Rescale the job if {@link Context#shouldRescale} is true. Otherwise, force a rescale using
+     * {@link Executing#forceRescale()} after {@link
+     * JobManagerOptions#SCHEDULER_SCALING_INTERVAL_MAX}.
+     */
     private void maybeRescale() {
-        if (context.shouldRescale(getExecutionGraph())) {
-            getLogger().info("Can change the parallelism of job. Restarting job.");
+        rescaleScheduled = false;
+        if (context.shouldRescale(getExecutionGraph(), false)) {
+            getLogger().info("Can change the parallelism of the job. Restarting the job.");
             context.goToRestarting(
                     getExecutionGraph(),
                     getExecutionGraphHandler(),
                     getOperatorCoordinatorHandler(),
                     Duration.ofMillis(0L),
                     getFailures());
+        } else if (scalingIntervalMax != null) {
+            getLogger()
+                    .info(
+                            "The longer the pipeline runs, the more the (small) resource gain is worth the restarting time. "
+                                    + "Last resource added does not meet {}, force a rescale after {} time({}) if the resource is still there.",
+                            JobManagerOptions.MIN_PARALLELISM_INCREASE,
+                            JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX.key(),
+                            scalingIntervalMax);
+            // schedule a force rescale in JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX time
+            context.runIfState(this, this::forceRescale, scalingIntervalMax);
+        }
```

Comment:
I still have a question for `scalingIntervalMax`.

> if minimum scaling requirements are not met but last rescale was done more than scaling-interval.max ago, a rescale is forced.

I copy it from FLIP, from it, we can see if last rescale was done more than scaling-interval.max ago, we should call forceRescale, right?

For current code, we didn't have this logic. We always schedule it after scalingIntervalMax. If the `timeSinceLastRescale()` > `scalingIntervalMax`, the delay isn't necessary, right?

Reply:
As you mentioned, we discussed a lot the `scalingIntervalMax` semantics and these semantics have changed during these discussions. I have not updated the FLIP to reflect these changes.
That being said, I agree I if `timeSinceLastRescale() > scalingIntervalMax` we can run an immediate force rescale. I'll update the FLIP as well

Reply:
done